File: go/internal/repository/analytics_repository.go
package repository

import (
	"time"

	"gorm.io/gorm"

	"saving_advisor_system_go/internal/models"
)

// AnalyticsRepository handles analytics data operations
type AnalyticsRepository struct {
	db *gorm.DB
}

// NewAnalyticsRepository creates a new analytics repository
func NewAnalyticsRepository(db *gorm.DB) *AnalyticsRepository {
	return &AnalyticsRepository{db: db}
}

// UserBehaviorAnalysis represents user behavior analytics
type UserBehaviorAnalysis struct {
	TotalTransactions      int64   `json:"total_transactions"`
	AvgTransactionValue    float64 `json:"avg_transaction_value"`
	TotalTransactionValue  float64 `json:"total_transaction_value"`
	BuyerTransactions      int64   `json:"buyer_transactions"`
	SellerTransactions     int64   `json:"seller_transactions"`
	AISessionCount         int64   `json:"ai_session_count"`
	FeedbackCount          int64   `json:"feedback_count"`
	AvgBalance             float64 `json:"avg_balance"`
	MaxBalance             float64 `json:"max_balance"`
	MinBalance             float64 `json:"min_balance"`
}

// GetUserBehaviorAnalysis retrieves user behavior analytics
func (r *AnalyticsRepository) GetUserBehaviorAnalysis(userID string) (*UserBehaviorAnalysis, error) {
	analysis := &UserBehaviorAnalysis{}

	// Transaction statistics
	var transactionStats struct {
		TotalTransactions     int64
		AvgTransactionValue   float64
		TotalTransactionValue float64
		BuyerTransactions     int64
		SellerTransactions    int64
	}

	err := r.db.Model(&models.Transaction{}).
		Select(`
			COUNT(*) as total_transactions,
			COALESCE(AVG(price), 0) as avg_transaction_value,
			COALESCE(SUM(price), 0) as total_transaction_value,
			SUM(CASE WHEN buyer_id = ? THEN 1 ELSE 0 END) as buyer_transactions,
			SUM(CASE WHEN seller_id = ? THEN 1 ELSE 0 END) as seller_transactions
		`, userID, userID).
		Where("buyer_id = ? OR seller_id = ?", userID, userID).
		Scan(&transactionStats).Error

	if err != nil {
		return nil, err
	}

	analysis.TotalTransactions = transactionStats.TotalTransactions
	analysis.AvgTransactionValue = transactionStats.AvgTransactionValue
	analysis.TotalTransactionValue = transactionStats.TotalTransactionValue
	analysis.BuyerTransactions = transactionStats.BuyerTransactions
	analysis.SellerTransactions = transactionStats.SellerTransactions

	// AI Session statistics
	thirtyDaysAgo := time.Now().AddDate(0, 0, -30)

	err = r.db.Model(&models.AISession{}).
		Where("user_id = ? AND created_at >= ?", userID, thirtyDaysAgo).
		Count(&analysis.AISessionCount).Error
	if err != nil {
		return nil, err
	}

	err = r.db.Model(&models.AISession{}).
		Where("user_id = ? AND created_at >= ? AND user_feedback IS NOT NULL", userID, thirtyDaysAgo).
		Count(&analysis.FeedbackCount).Error
	if err != nil {
		return nil, err
	}

	// Balance statistics
	var balanceStats struct {
		AvgBalance float64
		MaxBalance float64
		MinBalance float64
	}

	err = r.db.Model(&models.DailyUserBalance{}).
		Select(`
			COALESCE(AVG(total_balance), 0) as avg_balance,
			COALESCE(MAX(total_balance), 0) as max_balance,
			COALESCE(MIN(total_balance), 0) as min_balance
		`).
		Where("user_id = ? AND date >= ?", userID, thirtyDaysAgo).
		Scan(&balanceStats).Error

	if err == nil {
		analysis.AvgBalance = balanceStats.AvgBalance
		analysis.MaxBalance = balanceStats.MaxBalance
		analysis.MinBalance = balanceStats.MinBalance
	}

	return analysis, nil
}

// MarketTrend represents market trend data
type MarketTrend struct {
	Date             string  `json:"date"`
	TransactionCount int64   `json:"transaction_count"`
	AvgPrice         float64 `json:"avg_price"`
	TotalValue       float64 `json:"total_value"`
}

// GetMarketTrends retrieves market trend analytics
func (r *AnalyticsRepository) GetMarketTrends(days int) ([]MarketTrend, error) {
	startDate := time.Now().AddDate(0, 0, -days)

	var trends []MarketTrend
	err := r.db.Model(&models.Transaction{}).
		Select(`
			DATE(created) as date,
			COUNT(*) as transaction_count,
			COALESCE(AVG(price), 0) as avg_price,
			COALESCE(SUM(price), 0) as total_value
		`).
		Where("created >= ?", startDate).
		Group("DATE(created)").
		Order("DATE(created)").
		Scan(&trends).Error

	return trends, err
}

// RecommendationPerformance represents recommendation system performance
type RecommendationPerformance struct {
	TotalRecommendations int     `json:"total_recommendations"`
	Accepted             int     `json:"accepted"`
	Rejected             int     `json:"rejected"`
	Pending              int     `json:"pending"`
	AcceptanceRate       float64 `json:"acceptance_rate"`
	AnalysisPeriod       string  `json:"analysis_period"`
}

// GetRecommendationPerformance retrieves recommendation system performance
func (r *AnalyticsRepository) GetRecommendationPerformance() (*RecommendationPerformance, error) {
	thirtyDaysAgo := time.Now().AddDate(0, 0, -30)

	perf := &RecommendationPerformance{
		AnalysisPeriod: "30 days",
	}

	var stats struct {
		Total    int64
		Accepted int64
		Rejected int64
		Pending  int64
	}

	err := r.db.Model(&models.Recommendation{}).
		Select(`
			COUNT(*) as total,
			SUM(CASE WHEN status = 'accepted' THEN 1 ELSE 0 END) as accepted,
			SUM(CASE WHEN status = 'rejected' THEN 1 ELSE 0 END) as rejected,
			SUM(CASE WHEN status = 'pending' THEN 1 ELSE 0 END) as pending
		`).
		Where("created_at >= ?", thirtyDaysAgo).
		Scan(&stats).Error

	if err != nil {
		return nil, err
	}

	perf.TotalRecommendations = int(stats.Total)
	perf.Accepted = int(stats.Accepted)
	perf.Rejected = int(stats.Rejected)
	perf.Pending = int(stats.Pending)

	if stats.Total > 0 {
		perf.AcceptanceRate = float64(stats.Accepted) / float64(stats.Total) * 100
	}

	return perf, nil
}

// UserEngagementMetrics represents user engagement metrics
type UserEngagementMetrics struct {
	TotalUsers            int     `json:"total_users"`
	WeeklyActiveUsers     int     `json:"weekly_active_users"`
	MonthlyActiveUsers    int     `json:"monthly_active_users"`
	WeeklyEngagementRate  float64 `json:"weekly_engagement_rate"`
	MonthlyEngagementRate float64 `json:"monthly_engagement_rate"`
}

// GetUserEngagementMetrics retrieves user engagement metrics
func (r *AnalyticsRepository) GetUserEngagementMetrics() (*UserEngagementMetrics, error) {
	metrics := &UserEngagementMetrics{}

	// Total users
	var totalUsers int64
	err := r.db.Model(&models.User{}).Count(&totalUsers).Error
	if err != nil {
		return nil, err
	}
	metrics.TotalUsers = int(totalUsers)

	// Weekly active users
	weekAgo := time.Now().AddDate(0, 0, -7)
	var weeklyActive int64
	err = r.db.Model(&models.AISession{}).
		Where("created_at >= ?", weekAgo).
		Distinct("user_id").
		Count(&weeklyActive).Error
	if err != nil {
		return nil, err
	}
	metrics.WeeklyActiveUsers = int(weeklyActive)

	// Monthly active users
	monthAgo := time.Now().AddDate(0, -1, 0)
	var monthlyActive int64
	err = r.db.Model(&models.AISession{}).
		Where("created_at >= ?", monthAgo).
		Distinct("user_id").
		Count(&monthlyActive).Error
	if err != nil {
		return nil, err
	}
	metrics.MonthlyActiveUsers = int(monthlyActive)

	// Calculate engagement rates
	if totalUsers > 0 {
		metrics.WeeklyEngagementRate = float64(weeklyActive) / float64(totalUsers) * 100
		metrics.MonthlyEngagementRate = float64(monthlyActive) / float64(totalUsers) * 100
	}

	return metrics, nil
}

File: app/db/repositories/analytics_repository.py
"""
Analytics data repository
Replaces BigQuery analytics functionality
"""
from sqlalchemy import select, and_, func, case, or_
from sqlalchemy.ext.asyncio import AsyncSession
from typing import Dict, Any, List
from datetime import datetime, timedelta

from app.db.models import User, Transaction, AISession, Recommendation, DailyUserBalance


class AnalyticsRepository:
    """Analytics data repository"""
    
    def __init__(self, session: AsyncSession):
        self.session = session
    
    async def get_user_behavior_analysis(self, user_id: str) -> Dict[str, Any]:
        """User behavior analysis"""
        # Transaction statistics
        transaction_stats = await self.session.execute(
            select(
                func.count(Transaction.transaction_id).label('total_transactions'),
                func.avg(Transaction.price).label('avg_transaction_value'),
                func.sum(Transaction.price).label('total_transaction_value'),
                func.count(case((Transaction.buyer_id == user_id, 1))).label('buyer_transactions'),
                func.count(case((Transaction.seller_id == user_id, 1))).label('seller_transactions'),
            )
            .where(
                or_(
                    Transaction.buyer_id == user_id,
                    Transaction.seller_id == user_id
                )
            )
        )
        trans_stats = transaction_stats.one()
        
        # AI会话统计
        ai_stats = await self.session.execute(
            select(
                func.count(AISession.session_id).label('session_count'),
                func.count(case((AISession.user_feedback.isnot(None), 1))).label('feedback_count'),
            )
            .where(
                and_(
                    AISession.user_id == user_id,
                    AISession.created_at >= datetime.utcnow() - timedelta(days=30)
                )
            )
        )
        ai_data = ai_stats.one()
        
        # Asset statistics
        balance_stats = await self.session.execute(
            select(
                func.avg(DailyUserBalance.total_balance).label('avg_balance'),
                func.max(DailyUserBalance.total_balance).label('max_balance'),
                func.min(DailyUserBalance.total_balance).label('min_balance'),
            )
            .where(
                and_(
                    DailyUserBalance.user_id == user_id,
                    DailyUserBalance.date >= datetime.utcnow() - timedelta(days=30)
                )
            )
        )
        balance_data = balance_stats.one()
        
        return {
            'total_transactions': trans_stats.total_transactions or 0,
            'avg_transaction_value': float(trans_stats.avg_transaction_value or 0),
            'total_transaction_value': float(trans_stats.total_transaction_value or 0),
            'buyer_transactions': trans_stats.buyer_transactions or 0,
            'seller_transactions': trans_stats.seller_transactions or 0,
            'ai_session_count': ai_data.session_count or 0,
            'feedback_count': ai_data.feedback_count or 0,
            'avg_balance': float(balance_data.avg_balance or 0),
            'max_balance': float(balance_data.max_balance or 0),
            'min_balance': float(balance_data.min_balance or 0),
        }
    
    async def get_market_trends(self, days: int = 30) -> Dict[str, Any]:
        """Market trend analysis"""
        start_date = datetime.utcnow() - timedelta(days=days)
        
        # Statistics by date
        daily_stats = await self.session.execute(
            select(
                func.date(Transaction.created).label('date'),
                func.count(Transaction.transaction_id).label('transaction_count'),
                func.avg(Transaction.price).label('avg_price'),
                func.sum(Transaction.price).label('total_value'),
            )
            .where(Transaction.created >= start_date)
            .group_by(func.date(Transaction.created))
            .order_by(func.date(Transaction.created))
        )
        
        trends = []
        for row in daily_stats:
            trends.append({
                'date': row.date.isoformat() if hasattr(row.date, 'isoformat') else str(row.date),
                'transaction_count': row.transaction_count or 0,
                'avg_price': float(row.avg_price or 0),
                'total_value': float(row.total_value or 0),
            })
        
        return {
            'period_days': days,
            'trends': trends,
            'start_date': start_date.isoformat(),
            'end_date': datetime.utcnow().isoformat(),
        }
    
    async def get_recommendation_performance(self) -> Dict[str, Any]:
        """Recommendation system effectiveness analysis"""
        # Recommendation statistics
        rec_stats = await self.session.execute(
            select(
                func.count(Recommendation.recommendation_id).label('total_recommendations'),
                func.count(case((Recommendation.status == 'accepted', 1))).label('accepted'),
                func.count(case((Recommendation.status == 'rejected', 1))).label('rejected'),
                func.count(case((Recommendation.status == 'pending', 1))).label('pending'),
            )
            .where(
                Recommendation.created_at >= datetime.utcnow() - timedelta(days=30)
            )
        )
        stats = rec_stats.one()
        
        total = stats.total_recommendations or 0
        acceptance_rate = (stats.accepted / total * 100) if total > 0 else 0.0
        
        return {
            'total_recommendations': total,
            'accepted': stats.accepted or 0,
            'rejected': stats.rejected or 0,
            'pending': stats.pending or 0,
            'acceptance_rate': acceptance_rate,
            'analysis_period': '30 days',
        }
    
    async def get_user_engagement_metrics(self) -> Dict[str, Any]:
        """User engagement metrics"""
        # Active user statistics
        active_users = await self.session.execute(
            select(func.count(func.distinct(AISession.user_id)))
            .where(AISession.created_at >= datetime.utcnow() - timedelta(days=7))
        )
        weekly_active = active_users.scalar() or 0
        
        active_users_monthly = await self.session.execute(
            select(func.count(func.distinct(AISession.user_id)))
            .where(AISession.created_at >= datetime.utcnow() - timedelta(days=30))
        )
        monthly_active = active_users_monthly.scalar() or 0
        
        # Total user count
        total_users = await self.session.execute(
            select(func.count(User.user_id))
        )
        total = total_users.scalar() or 0
        
        return {
            'total_users': total,
            'weekly_active_users': weekly_active,
            'monthly_active_users': monthly_active,
            'weekly_engagement_rate': (weekly_active / total * 100) if total > 0 else 0.0,
            'monthly_engagement_rate': (monthly_active / total * 100) if total > 0 else 0.0,
        }

